File: spark/so/grpc_test/swap_test.go
package grpctest

import (
	"bytes"
	"context"
	"testing"
	"time"

	"github.com/btcsuite/btcd/btcec/v2"
	"github.com/btcsuite/btcd/txscript"
	"github.com/decred/dcrd/dcrec/secp256k1/v4"
	"github.com/lightsparkdev/spark/common"
	"github.com/lightsparkdev/spark/proto/spark"
	testutil "github.com/lightsparkdev/spark/test_util"
	"github.com/lightsparkdev/spark/wallet"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestSwap(t *testing.T) {
	// Initiate sender
	senderConfig, err := testutil.TestWalletConfig()
	if err != nil {
		t.Fatalf("failed to create sender wallet config: %v", err)
	}

	senderLeafPrivKey, err := secp256k1.GeneratePrivateKey()
	if err != nil {
		t.Fatalf("failed to create node signing private key: %v", err)
	}
	senderRootNode, err := testutil.CreateNewTree(senderConfig, faucet, senderLeafPrivKey, 100_000)
	if err != nil {
		t.Fatalf("failed to create new tree: %v", err)
	}

	// Initiate receiver
	receiverConfig, err := testutil.TestWalletConfig()
	if err != nil {
		t.Fatalf("failed to create sender wallet config: %v", err)
	}

	receiverLeafPrivKey, err := secp256k1.GeneratePrivateKey()
	if err != nil {
		t.Fatalf("failed to create node signing private key: %v", err)
	}
	receiverRootNode, err := testutil.CreateNewTree(receiverConfig, faucet, receiverLeafPrivKey, 100_000)
	if err != nil {
		t.Fatalf("failed to create new tree: %v", err)
	}

	// Sender initiates transfer
	senderNewLeafPrivKey, err := secp256k1.GeneratePrivateKey()
	if err != nil {
		t.Fatalf("failed to create new node signing private key: %v", err)
	}

	senderTransferNode := wallet.LeafKeyTweak{
		Leaf:              senderRootNode,
		SigningPrivKey:    senderLeafPrivKey.Serialize(),
		NewSigningPrivKey: senderNewLeafPrivKey.Serialize(),
	}
	senderLeavesToTransfer := [1]wallet.LeafKeyTweak{senderTransferNode}

	// Get signature for refunds (normal flow)
	senderTransfer, senderRefundSignatureMap, leafDataMap, err := wallet.StartSwapSignRefund(
		context.Background(),
		senderConfig,
		senderLeavesToTransfer[:],
		receiverConfig.IdentityPublicKey(),
		time.Now().Add(10*time.Minute),
	)
	assert.NoError(t, err)
	assert.Equal(t, 1, len(senderRefundSignatureMap), "expected 1 refund signature")
	signature := senderRefundSignatureMap[senderRootNode.Id]
	assert.NotNil(t, signature, "expected refund signature for root node")
	leafData := leafDataMap[senderRootNode.Id]
	assert.NotNil(t, leafData, "expected leaf data for root node")

	sighash, err := common.SigHashFromTx(leafData.RefundTx, 0, leafData.Tx.TxOut[leafData.Vout])
	assert.NoError(t, err)

	// Create adaptor from that signature
	adaptorAddedSignature, adaptorPrivKey, err := common.GenerateAdaptorFromSignature(signature)
	assert.NoError(t, err)
	_, adaptorPub := btcec.PrivKeyFromBytes(adaptorPrivKey)

	// Alice sends adaptor and signature to Bob, Bob validates the adaptor
	nodeVerifyingPubkey, err := secp256k1.ParsePubKey(senderRootNode.VerifyingPublicKey)
	assert.NoError(t, err)
	taprootKey := txscript.ComputeTaprootKeyNoScript(nodeVerifyingPubkey)
	err = common.ValidateOutboundAdaptorSignature(taprootKey, sighash, adaptorAddedSignature, adaptorPub.SerializeCompressed())
	assert.NoError(t, err)

	// Bob signs refunds with adaptor
	receiverNewLeafPrivKey, err := secp256k1.GeneratePrivateKey()
	assert.NoError(t, err)

	receiverTransferNode := wallet.LeafKeyTweak{
		Leaf:              receiverRootNode,
		SigningPrivKey:    receiverLeafPrivKey.Serialize(),
		NewSigningPrivKey: receiverNewLeafPrivKey.Serialize(),
	}
	receiverLeavesToTransfer := [1]wallet.LeafKeyTweak{receiverTransferNode}
	receiverTransfer, receiverRefundSignatureMap, leafDataMap, operatorSigningResults, err := wallet.CounterSwapSignRefund(
		context.Background(),
		receiverConfig,
		receiverLeavesToTransfer[:],
		senderConfig.IdentityPublicKey(),
		time.Now().Add(10*time.Minute),
		adaptorPub,
	)
	assert.NoError(t, err)

	// Alice verifies Bob's signatures
	receiverSighash, err := common.SigHashFromTx(leafDataMap[receiverLeavesToTransfer[0].Leaf.Id].RefundTx, 0, leafDataMap[receiverLeavesToTransfer[0].Leaf.Id].Tx.TxOut[leafDataMap[receiverLeavesToTransfer[0].Leaf.Id].Vout])
	assert.NoError(t, err)

	receiverKey, err := secp256k1.ParsePubKey(receiverLeavesToTransfer[0].Leaf.VerifyingPublicKey)
	assert.NoError(t, err)
	receiverTaprootKey := txscript.ComputeTaprootKeyNoScript(receiverKey)

	_, err = common.ApplyAdaptorToSignature(receiverTaprootKey, receiverSighash, receiverRefundSignatureMap[receiverLeavesToTransfer[0].Leaf.Id], adaptorPrivKey)
	assert.NoError(t, err)

	// Alice reveals adaptor secret to Bob, Bob combines with existing adaptor signatures to get valid signatures
	newReceiverRefundSignatureMap := make(map[string][]byte)
	for nodeID, signature := range receiverRefundSignatureMap {
		leafData := leafDataMap[nodeID]
		sighash, _ := common.SigHashFromTx(leafData.RefundTx, 0, leafData.Tx.TxOut[leafData.Vout])
		var verifyingPubkey *secp256k1.PublicKey
		for _, signingResult := range operatorSigningResults {
			if signingResult.LeafId == nodeID {
				verifyingPubkey, err = secp256k1.ParsePubKey(signingResult.VerifyingKey)
				assert.NoError(t, err)
			}
		}
		assert.NotNil(t, verifyingPubkey, "expected signing result for leaf %s", nodeID)
		taprootKey := txscript.ComputeTaprootKeyNoScript(verifyingPubkey)
		adaptorSig, err := common.ApplyAdaptorToSignature(taprootKey, sighash, signature, adaptorPrivKey)
		assert.NoError(t, err)
		newReceiverRefundSignatureMap[nodeID] = adaptorSig
	}

	// Alice provides key tweak, Bob claims alice's leaves
	senderTransfer, err = wallet.SendTransferTweakKey(
		context.Background(),
		senderConfig,
		senderTransfer,
		senderLeavesToTransfer[:],
		senderRefundSignatureMap,
	)
	if err != nil {
		t.Fatalf("failed to send transfer tweak key: %v", err)
	}

	receiverToken, err := wallet.AuthenticateWithServer(context.Background(), receiverConfig)
	if err != nil {
		t.Fatalf("failed to authenticate receiver: %v", err)
	}
	receiverCtx := wallet.ContextWithToken(context.Background(), receiverToken)
	pendingTransfer, err := wallet.QueryPendingTransfers(receiverCtx, receiverConfig)
	if err != nil {
		t.Fatalf("failed to query pending transfers: %v", err)
	}
	if len(pendingTransfer.Transfers) != 1 {
		t.Fatalf("expected 1 pending transfer, got %d", len(pendingTransfer.Transfers))
	}
	receiverPendingTransfer := pendingTransfer.Transfers[0]
	if receiverPendingTransfer.Id != senderTransfer.Id {
		t.Fatalf("expected transfer id %s, got %s", senderTransfer.Id, receiverPendingTransfer.Id)
	}
	require.Equal(t, receiverPendingTransfer.Type, spark.TransferType_SWAP)

	leafPrivKeyMap, err := wallet.VerifyPendingTransfer(context.Background(), receiverConfig, receiverPendingTransfer)
	if err != nil {
		t.Fatalf("unable to verify pending transfer: %v", err)
	}
	if len(*leafPrivKeyMap) != 1 {
		t.Fatalf("Expected 1 leaf to transfer, got %d", len(*leafPrivKeyMap))
	}
	if !bytes.Equal((*leafPrivKeyMap)[senderRootNode.Id], senderNewLeafPrivKey.Serialize()) {
		t.Fatalf("wrong leaf signing private key")
	}

	finalLeafPrivKey, err := secp256k1.GeneratePrivateKey()
	if err != nil {
		t.Fatalf("failed to create new node signing private key: %v", err)
	}
	claimingNode := wallet.LeafKeyTweak{
		Leaf:              receiverPendingTransfer.Leaves[0].Leaf,
		SigningPrivKey:    senderNewLeafPrivKey.Serialize(),
		NewSigningPrivKey: finalLeafPrivKey.Serialize(),
	}
	leavesToClaim := [1]wallet.LeafKeyTweak{claimingNode}
	_, err = wallet.ClaimTransfer(
		receiverCtx,
		receiverPendingTransfer,
		receiverConfig,
		leavesToClaim[:],
	)
	if err != nil {
		t.Fatalf("failed to ClaimTransfer: %v", err)
	}

	// Bob provides key tweak, Alice claims bob's leaves
	_, err = wallet.SendTransferTweakKey(
		context.Background(),
		receiverConfig,
		receiverTransfer,
		receiverLeavesToTransfer[:],
		newReceiverRefundSignatureMap,
	)
	if err != nil {
		t.Fatalf("failed to send transfer tweak key: %v", err)
	}

	senderToken, err := wallet.AuthenticateWithServer(context.Background(), senderConfig)
	if err != nil {
		t.Fatalf("failed to authenticate receiver: %v", err)
	}
	senderCtx := wallet.ContextWithToken(context.Background(), senderToken)
	pendingTransfer, err = wallet.QueryPendingTransfers(senderCtx, senderConfig)
	if err != nil {
		t.Fatalf("failed to query pending transfers: %v", err)
	}
	if len(pendingTransfer.Transfers) != 1 {
		t.Fatalf("expected 1 pending transfer, got %d", len(pendingTransfer.Transfers))
	}
	senderPendingTransfer := pendingTransfer.Transfers[0]
	if receiverPendingTransfer.Id != senderTransfer.Id {
		t.Fatalf("expected transfer id %s, got %s", senderTransfer.Id, receiverPendingTransfer.Id)
	}
	require.Equal(t, senderPendingTransfer.Type, spark.TransferType_COUNTER_SWAP)

	leafPrivKeyMap, err = wallet.VerifyPendingTransfer(context.Background(), senderConfig, senderPendingTransfer)
	if err != nil {
		t.Fatalf("unable to verify pending transfer: %v", err)
	}
	if len(*leafPrivKeyMap) != 1 {
		t.Fatalf("Expected 1 leaf to transfer, got %d", len(*leafPrivKeyMap))
	}
	if !bytes.Equal((*leafPrivKeyMap)[receiverRootNode.Id], receiverNewLeafPrivKey.Serialize()) {
		t.Fatalf("wrong leaf signing private key")
	}

	finalLeafPrivKey, err = secp256k1.GeneratePrivateKey()
	if err != nil {
		t.Fatalf("failed to create new node signing private key: %v", err)
	}
	claimingNode = wallet.LeafKeyTweak{
		Leaf:              senderPendingTransfer.Leaves[0].Leaf,
		SigningPrivKey:    receiverNewLeafPrivKey.Serialize(),
		NewSigningPrivKey: finalLeafPrivKey.Serialize(),
	}
	leavesToClaim = [1]wallet.LeafKeyTweak{claimingNode}
	_, err = wallet.ClaimTransfer(
		senderCtx,
		senderPendingTransfer,
		senderConfig,
		leavesToClaim[:],
	)
	if err != nil {
		t.Fatalf("failed to ClaimTransfer: %v", err)
	}
}

File: sdks/js/packages/spark-sdk/src/tests/integration/swap.test.ts
import { describe, expect, it } from "@jest/globals";
import { equalBytes, hexToBytes } from "@noble/curves/abstract/utils";
import { secp256k1 } from "@noble/curves/secp256k1";
import { sha256 } from "@scure/btc-signer/utils";
import { WalletConfigService } from "../../services/config.js";
import { ConnectionManager } from "../../services/connection.js";
import { LeafKeyTweak, TransferService } from "../../services/transfer.js";
import {
  applyAdaptorToSignature,
  generateAdaptorFromSignature,
} from "../../utils/adaptor-signature.js";
import {
  computeTaprootKeyNoScript,
  getSigHashFromTx,
} from "../../utils/bitcoin.js";
import { createNewTree } from "../test-util.js";
import { SparkWalletTesting } from "../utils/spark-testing-wallet.js";
import { BitcoinFaucet } from "../utils/test-faucet.js";
import { ValidationError } from "../../errors/types.js";

describe("swap", () => {
  it("test swap", async () => {
    const faucet = BitcoinFaucet.getInstance();
    // Initiate sender
    const { wallet: senderWallet } = await SparkWalletTesting.initialize({
      options: {
        network: "LOCAL",
      },
    });
    const senderPubkey = await senderWallet.getIdentityPublicKey();

    const senderConfig = new WalletConfigService(
      {
        network: "LOCAL",
      },
      senderWallet.getSigner(),
    );
    const senderConnectionManager = new ConnectionManager(senderConfig);
    const senderTransferService = new TransferService(
      senderConfig,
      senderConnectionManager,
    );

    // Initiate receiver
    const { wallet: receiverWallet } = await SparkWalletTesting.initialize({
      options: {
        network: "LOCAL",
      },
    });
    const receiverPubkey = await receiverWallet.getIdentityPublicKey();

    const receiverConfig = new WalletConfigService(
      {
        network: "LOCAL",
      },
      receiverWallet.getSigner(),
    );
    const receiverConnectionManager = new ConnectionManager(receiverConfig);
    const receiverTransferService = new TransferService(
      receiverConfig,
      receiverConnectionManager,
    );

    const senderLeafPubKey = await senderWallet.getSigner().generatePublicKey();
    const senderRootNode = await createNewTree(
      senderWallet,
      senderLeafPubKey,
      faucet,
    );

    const receiverLeafPubKey = await receiverWallet
      .getSigner()
      .generatePublicKey();
    const receiverRootNode = await createNewTree(
      receiverWallet,
      receiverLeafPubKey,
      faucet,
    );

    // Sender initiates transfer
    const senderNewLeafPubKey = await senderWallet
      .getSigner()
      .generatePublicKey(sha256("1"));
    const senderTransferNode: LeafKeyTweak = {
      leaf: senderRootNode,
      signingPubKey: senderLeafPubKey,
      newSigningPubKey: senderNewLeafPubKey,
    };
    const senderLeavesToTransfer = [senderTransferNode];

    // Get signature for refunds (normal flow)
    const {
      transfer: senderTransfer,
      signatureMap: senderRefundSignatureMap,
      leafDataMap: senderLeafDataMap,
    } = await senderTransferService.startSwapSignRefund(
      senderLeavesToTransfer,
      hexToBytes(receiverPubkey),
      new Date(Date.now() + 10 * 60 * 1000),
    );

    expect(senderRefundSignatureMap.size).toBe(1);
    const senderSignature = senderRefundSignatureMap.get(senderRootNode.id);
    expect(senderSignature).toBeDefined();
    expect(senderLeafDataMap.size).toBe(1);

    const { adaptorPrivateKey, adaptorSignature } =
      generateAdaptorFromSignature(senderSignature!);
    const adaptorPubKey = secp256k1.getPublicKey(adaptorPrivateKey);

    const receiverNewLeafPubKey = await receiverWallet
      .getSigner()
      .generatePublicKey(sha256("1"));

    const receiverTransferNode: LeafKeyTweak = {
      leaf: receiverRootNode,
      signingPubKey: receiverLeafPubKey,
      newSigningPubKey: receiverNewLeafPubKey,
    };
    const receiverLeavesToTransfer = [receiverTransferNode];

    const {
      transfer: receiverTransfer,
      signatureMap: receiverRefundSignatureMap,
      leafDataMap: receiverLeafDataMap,
      signingResults: operatorSigningResults,
    } = await receiverTransferService.counterSwapSignRefund(
      receiverLeavesToTransfer,
      hexToBytes(senderPubkey),
      new Date(Date.now() + 10 * 60 * 1000),
      adaptorPubKey,
    );

    const newReceiverRefundSignatureMap = new Map<string, Uint8Array>();
    for (const [nodeId, signature] of receiverRefundSignatureMap.entries()) {
      const leafData = receiverLeafDataMap.get(nodeId);
      if (!leafData?.refundTx) {
        throw new ValidationError("Refund transaction not found", {
          field: "refundTx",
          value: leafData,
        });
      }
      const sighash = getSigHashFromTx(
        leafData.refundTx,
        0,
        leafData.tx.getOutput(leafData.vout),
      );
      let verifyingPubkey: Uint8Array | undefined;
      for (const signingResult of operatorSigningResults) {
        if (signingResult.leafId === nodeId) {
          verifyingPubkey = signingResult.verifyingKey;
        }
      }
      expect(verifyingPubkey).toBeDefined();
      const taprootKey = computeTaprootKeyNoScript(
        verifyingPubkey!.slice(1, 33),
      );
      const adaptorSig = applyAdaptorToSignature(
        taprootKey.slice(1, 33),
        sighash,
        signature,
        adaptorPrivateKey,
      );
      newReceiverRefundSignatureMap.set(nodeId, adaptorSig);
    }
    const senderTransferTweakKey =
      await senderTransferService.sendTransferTweakKey(
        senderTransfer,
        senderLeavesToTransfer,
        senderRefundSignatureMap,
      );

    const pendingTransfer =
      await receiverTransferService.queryPendingTransfers();
    expect(pendingTransfer.transfers.length).toBe(1);
    const receiverPendingTransfer = pendingTransfer.transfers[0];
    expect(receiverPendingTransfer!.id).toBe(senderTransferTweakKey.id);

    const leafPrivKeyMap = await receiverTransferService.verifyPendingTransfer(
      receiverPendingTransfer!,
    );

    expect(leafPrivKeyMap.size).toBe(1);
    expect(leafPrivKeyMap.get(senderRootNode.id)).toBeDefined();
    const bytesEqual = equalBytes(
      leafPrivKeyMap.get(senderRootNode.id)!,
      senderNewLeafPubKey,
    );
    expect(bytesEqual).toBe(true);
    expect(receiverPendingTransfer!.leaves[0]!.leaf).toBeDefined();
    const finalLeafPubKey = await receiverWallet
      .getSigner()
      .generatePublicKey(sha256("2"));
    const claimingNode: LeafKeyTweak = {
      leaf: receiverPendingTransfer!.leaves[0]!.leaf!,
      signingPubKey: senderNewLeafPubKey,
      newSigningPubKey: finalLeafPubKey,
    };
    const leavesToClaim = [claimingNode];
    await receiverTransferService.claimTransfer(
      receiverPendingTransfer!,
      leavesToClaim,
    );
    await receiverTransferService.sendTransferTweakKey(
      receiverTransfer,
      receiverLeavesToTransfer,
      newReceiverRefundSignatureMap,
    );

    const sPendingTransfer =
      await senderTransferService.queryPendingTransfers();
    expect(sPendingTransfer.transfers.length).toBe(1);
    const senderPendingTransfer = sPendingTransfer.transfers[0];
    expect(senderPendingTransfer!.id).toBe(receiverTransfer.id);

    const senderLeafPrivKeyMap =
      await senderTransferService.verifyPendingTransfer(senderPendingTransfer!);
    expect(senderLeafPrivKeyMap.size).toBe(1);
    expect(senderLeafPrivKeyMap.get(receiverRootNode.id)).toBeDefined();
    const bytesEqual_1 = equalBytes(
      senderLeafPrivKeyMap.get(receiverRootNode.id)!,
      receiverNewLeafPubKey,
    );
    expect(bytesEqual_1).toBe(true);
    expect(senderPendingTransfer!.leaves[0]!.leaf).toBeDefined();

    const finalLeafPubKey_1 = await senderWallet
      .getSigner()
      .generatePublicKey(sha256("3"));
    const claimingNode_1: LeafKeyTweak = {
      leaf: senderPendingTransfer!.leaves[0]!.leaf!,
      signingPubKey: receiverNewLeafPubKey,
      newSigningPubKey: finalLeafPubKey_1,
    };
    const leavesToClaim_1 = [claimingNode_1];
    await senderTransferService.claimTransfer(
      senderPendingTransfer!,
      leavesToClaim_1,
    );
  }, 30000);
});
